topdown/crypto.go

```diff
@@ -32,67 +32,71 @@ const (
 )
 
 func builtinCryptoX509ParseCertificates(a ast.Value) (ast.Value, error) {
-
 	input, err := builtins.StringOperand(a, 1)
 	if err != nil {
 		return nil, err
 	}
 
-	// data to be passed to x509.ParseCertificates
-	bytes := []byte(input)
+	certs, err := getX509CertsFromBytes([]byte(input))
+	if err != nil {
+		return nil, err
+	}
 
-	// if the input is not a PEM string, attempt to decode b64
-	if str := string(input); !strings.HasPrefix(str, "-----BEGIN CERTIFICATE-----") {
-		bytes, err = base64.StdEncoding.DecodeString(str)
-		if err != nil {
-			return nil, err
-		}
+	return marshalX509Certs(certs)
+}
+
+func builtinCryptoX509ParseAndVerifyCertificates(
+	bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term) error) error {
+
+	a := args[0].Value
+	input, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return err
 	}
 
-	// attempt to decode input as PEM data
-	p, rest := pem.Decode(bytes)
-	if p != nil && p.Type != blockTypeCertificate {
-		return nil, fmt.Errorf("PEM data contains '%s', expected CERTIFICATE", p.Type)
+	certs, err := getX509CertsFromBytes([]byte(input))
+	if err != nil {
+		return err
 	}
-	if p != nil {
-		// if PEM decoded as a valid certificate, use its data as the DER input
-		bytes = p.Bytes
+
+	if len(certs) < 2 {
+		return fmt.Errorf("must supply at least two certificates to verify the chain")
```

Comment:
❓ I suppose this would be an _operand error_? We've got a type for that, I believe.

Reply:
Thanks, I wasn't aware of that.